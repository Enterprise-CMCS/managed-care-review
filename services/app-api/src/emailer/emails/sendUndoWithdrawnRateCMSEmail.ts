import type {
    EmailConfiguration,
    EmailData,
    StateAnalystsEmails,
} from '../emailer'
import type { ProgramType, RateType } from '../../domain-models'
import {
    validateAndParseUnwithdrawnRate,
    type undoWithdrawnRateEtaData,
} from './sendUndoWithdrawnRateStateEmail'
import { pruneDuplicateEmails } from '../formatters'
import { formatCalendarDate } from '@mc-review/dates'
import { renderTemplate, stripHTMLFromTemplate } from '../templateHelpers'

export const sendUndoWithdrawnRateCMSEmail = async (
    rate: RateType,
    statePrograms: ProgramType[],
    stateAnalystsEmails: StateAnalystsEmails,
    config: EmailConfiguration
): Promise<EmailData | Error> => {
    const undoWithdrawnRateData = validateAndParseUnwithdrawnRate(
        rate,
        statePrograms,
        config
    )

    if (undoWithdrawnRateData instanceof Error) {
        return undoWithdrawnRateData
    }

    const {
        latestStatusAction, //Contains update information
        rateInfo, //Rate name
        associatedContracts, //Contracts connected to the rate
    } = undoWithdrawnRateData

    const toAddresses = pruneDuplicateEmails([
        ...stateAnalystsEmails,
        ...config.dmcpSubmissionEmails,
        ...config.oactEmails,
        ...config.devReviewTeamEmails,
    ])

    const etaData: undoWithdrawnRateEtaData = {
        rateInfo: rateInfo,
        updatedBy: latestStatusAction.updatedBy.email,
        updatedOn: formatCalendarDate(
            latestStatusAction.updatedAt,
            'America/Los_Angeles'
        ),
        reason: latestStatusAction.updatedReason,
        associatedContracts: associatedContracts,
    }

    const template = await renderTemplate<undoWithdrawnRateEtaData>(
        'sendUndoWithdrawnRateCMSEmail',
        etaData
    )

    if (template instanceof Error) {
        return template
    } else {
        return {
            toAddresses,
            replyToAddresses: [],
            sourceEmail: config.emailSource,
            subject: `${
                config.stage !== 'prod' ? `[${config.stage}] ` : ''
            }${etaData.rateInfo.rateName} status update`,
            bodyText: stripHTMLFromTemplate(template),
            bodyHTML: template,
        }
    }
}
