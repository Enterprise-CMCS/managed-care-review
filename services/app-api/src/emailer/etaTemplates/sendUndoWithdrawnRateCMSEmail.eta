<h1 style=<%~'"font-weight: normal;"'%>>Rate <%= it.rateName %> status was updated to 'Submitted' by CMS</h1>
<b>Updated by:</b> <%= it.updatedBy %><br />
<b>Updated on:</b> <%= it.updatedOn %><br />
<b>Reason:</b> <%= it.reason %><br />
<b>Related contract action(s):</b>
<% if (it.associatedContracts.length > 1) { %>
<ul>
<% it.associatedContracts.forEach(function(contract){ %>
  <li><a href="<%= contract.summaryURL %>"><%= contract.contractName %></a></li>
<% }) %>
</ul>
<% } else { %>
<a href="<%= it.associatedContracts[0].summaryURL %>"><%= it.associatedContracts[0].contractName %></a><br />
<br />
<% } %>