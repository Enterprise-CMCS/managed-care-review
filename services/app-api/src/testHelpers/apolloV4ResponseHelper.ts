import type { GraphQLFormattedError } from 'graphql'
import type { ApolloServer } from '@apollo/server'

export type TestGraphQLResponse<T = any> = {
    data?: T
    errors?: GraphQLFormattedError[]
}

export function extractGraphQLResponse<T = any>(
    response: any
): TestGraphQLResponse<T> {
    if ('body' in response && response.body) {
        const result =
            response.body.kind === 'single'
                ? response.body.singleResult
                : response.body

        return {
            data: result.data,
            errors: result.errors,
        }
    }

    return {
        data: response.data,
        errors: response.errors,
    }
}

export async function executeGraphQLOperation<TData = any, TVariables = any>(
    server: ApolloServer,
    options: {
        query: any
        variables?: TVariables
        contextValue?: any
    }
): Promise<TestGraphQLResponse<TData>> {
    const response = await server.executeOperation({
        query: options.query,
        variables: options.variables as any,
    })

    return extractGraphQLResponse<TData>(response)
}
