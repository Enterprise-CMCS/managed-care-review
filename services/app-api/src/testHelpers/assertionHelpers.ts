// Functions that assists with type narrow in jest tests

// must -  interrupts test flow to through an error
function must<T>(maybeErr: T | Error): T {
    if (maybeErr instanceof Error) {
        throw maybeErr
    }
    return maybeErr
}

// expectToBeDefined - properly  type narrows the results of toBeDefined
// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/41179 would eventually address this but is not implemented
function expectToBeDefined<T>(arg: T): asserts arg is NonNullable<T> {
    expect(arg).toBeDefined()
}

export { must, expectToBeDefined }
