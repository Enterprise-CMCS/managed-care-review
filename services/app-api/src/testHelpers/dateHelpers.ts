// returns today's date in the ISO standard form. like '2022-02-24'
function todaysDate(): string {
    const today = new Date()
    return today.toISOString().split('T')[0]
}

export { todaysDate }
