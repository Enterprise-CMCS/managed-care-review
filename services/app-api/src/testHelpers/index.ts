export {
    assertAnError,
    assertAnErrorExtensions,
    assertAnErrorCode,
} from './gqlAssertions'

export { must } from './assertionHelpers'

export {
    mockInsertContractArgs,
    mockContractRevision,
} from './contractDataMocks'

export {
    mockInsertRateArgs,
    mockRateRevision,
    mockDraftRate,
} from './rateDataMocks'

export { testS3Client } from './s3Helpers'
export { getStateRecord } from './stateHelpers'

export { consoleLogFullData } from './debugHelpers'

export {
    fetchTestRateById,
    submitTestRate,
    unlockTestRate,
    updateTestRate,
} from './gqlRateHelpers'

export {
    createAndSubmitTestContract,
    fetchTestContract,
    approveTestContract,
    updateTestContractDraftRevision,
    createTestContract,
    fetchTestContractWithQuestions,
} from './gqlContractHelpers'

export {
    clearDocMetadata,
    clearMetadataFromContractFormData,
    clearMetadataFromRateFormData,
} from './documentHelpers'

export { mockGqlContractDraftRevisionFormDataInput } from './gqlContractInputMocks'
