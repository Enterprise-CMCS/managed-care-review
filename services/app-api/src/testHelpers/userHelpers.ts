import { NewPostgresStore, type InsertUserArgsType } from '../postgres'
import type {
    AdminUserType,
    CMSApproverUserType,
    CMSUserType,
    StateUserType,
    UserType,
    BusinessOwnerUserType,
    HelpdeskUserType,
} from '../domain-models'
import { sharedTestPrismaClient } from './storeHelpers'
import { v4 as uuidv4 } from 'uuid'

const testCMSUser = (userData?: Partial<CMSUserType>): CMSUserType => ({
    id: uuidv4(),
    role: 'CMS_USER',
    email: 'zuko@example.com',
    familyName: 'Zuko',
    givenName: 'Prince',
    divisionAssignment: 'DMCO' as const,
    stateAssignments: [],
    ...userData,
})

const testCMSApproverUser = (
    userData?: Partial<CMSApproverUserType>
): CMSApproverUserType => ({
    id: uuidv4(),
    role: 'CMS_APPROVER_USER',
    email: 'azula@example.com',
    familyName: 'Azula',
    givenName: 'Princess',
    divisionAssignment: 'DMCO' as const,
    stateAssignments: [],
    ...userData,
})

const testStateUser = (userData?: Partial<StateUserType>): StateUserType => ({
    id: uuidv4(),
    stateCode: 'FL',
    role: 'STATE_USER',
    email: 'james@example.com',
    familyName: 'Brown',
    givenName: 'James',
    ...userData,
})

const testAdminUser = (userData?: Partial<AdminUserType>): AdminUserType => ({
    id: uuidv4(),
    role: 'ADMIN_USER',
    email: 'iroh@example.com',
    familyName: 'Iroh',
    givenName: 'Uncle',
    ...userData,
})

const testBusinessOwnerUser = (
    userData?: Partial<BusinessOwnerUserType>
): BusinessOwnerUserType => ({
    id: uuidv4(),
    role: 'BUSINESSOWNER_USER',
    email: 'iroh@example.com',
    familyName: 'Iroh',
    givenName: 'Uncle',
    ...userData,
})

const testHelpdeskUser = (
    userData?: Partial<HelpdeskUserType>
): HelpdeskUserType => ({
    id: uuidv4(),
    role: 'HELPDESK_USER',
    email: 'iroh@example.com',
    familyName: 'Iroh',
    givenName: 'Uncle',
    ...userData,
})

const createDBUsersWithFullData = async (
    users: UserType[]
): Promise<UserType[]> => {
    const prismaClient = await sharedTestPrismaClient()
    const postgresStore = NewPostgresStore(prismaClient)

    const usersSeed: InsertUserArgsType[] = users.map((user: UserType) => {
        const userArgs: InsertUserArgsType & { id?: string } = {
            ...user,
            userID: user.id,
        }

        delete userArgs.id

        return userArgs as InsertUserArgsType
    })

    const result = await postgresStore.insertManyUsers(usersSeed)

    if (result instanceof Error) {
        throw result
    }

    return result
}

const iterableCmsUsersMockData: {
    userRole: 'CMS_USER' | 'CMS_APPROVER_USER'
    mockUser: <T>(userData?: Partial<T>) => CMSUserType | CMSApproverUserType
}[] = [
    {
        userRole: 'CMS_USER',
        mockUser: testCMSUser,
    },
    {
        userRole: 'CMS_APPROVER_USER',
        mockUser: testCMSApproverUser,
    },
]

const iterableNonCMSUsersMockData: {
    userRole:
        | 'HELPDESK_USER'
        | 'BUSINESSOWNER_USER'
        | 'ADMIN_USER'
        | 'STATE_USER'
    mockUser: <T>(
        userData?: Partial<T>
    ) =>
        | AdminUserType
        | BusinessOwnerUserType
        | HelpdeskUserType
        | StateUserType
}[] = [
    {
        userRole: 'ADMIN_USER',
        mockUser: testAdminUser,
    },
    {
        userRole: 'BUSINESSOWNER_USER',
        mockUser: testBusinessOwnerUser,
    },
    {
        userRole: 'HELPDESK_USER',
        mockUser: testHelpdeskUser,
    },
    {
        userRole: 'STATE_USER',
        mockUser: testStateUser,
    },
]

export {
    testAdminUser,
    testStateUser,
    testCMSUser,
    createDBUsersWithFullData,
    iterableCmsUsersMockData,
    iterableNonCMSUsersMockData,
    testCMSApproverUser,
    testBusinessOwnerUser,
    testHelpdeskUser,
}
