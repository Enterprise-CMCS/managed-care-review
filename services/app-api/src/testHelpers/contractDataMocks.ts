import { v4 as uuidv4 } from 'uuid'
import type {
    ContractRevisionTableWithRates,
    ContractTableFullPayload,
} from '../postgres/contractAndRates/prismaFullContractRateHelpers'
import type { StateCodeType } from '@mc-review/submissions'
import type { ContractFormDataType } from '../domain-models'
import { findStatePrograms, type InsertContractArgsType } from '../postgres'
import { must } from './assertionHelpers'

const defaultContractData = () => ({
    id: uuidv4(),
    createdAt: new Date(),
    updatedAt: new Date(),
    mccrsID: null,
    stateCode: 'MN',
    stateNumber: 111,
})

const mockInsertContractArgs = ({
    stateCode = 'MN',
    ...formData
}: {
    stateCode?: StateCodeType
} & Partial<ContractFormDataType>): InsertContractArgsType => {
    const statePrograms = must(findStatePrograms(stateCode))

    return {
        stateCode: stateCode,
        submissionType: formData?.submissionType ?? 'CONTRACT_AND_RATES',
        submissionDescription:
            formData?.submissionDescription ?? 'Contract 1.0',
        contractType: formData?.contractType ?? 'BASE',
        programIDs: formData?.programIDs ?? [statePrograms[0].id],
        populationCovered: formData?.populationCovered ?? 'MEDICAID',
        riskBasedContract: formData?.riskBasedContract ?? false,
        ...formData,
    }
}

const mockContractRevision = (
    contract?: Partial<ContractTableFullPayload>,
    revision?: Partial<ContractRevisionTableWithRates>,
    stateCode: StateCodeType = 'MN'
): ContractRevisionTableWithRates => {
    const statePrograms = must(findStatePrograms(stateCode))

    return {
        id: uuidv4(),
        contract: {
            ...defaultContractData(),
            ...contract,
        },
        relatedSubmisions: [],
        createdAt: new Date(),
        updatedAt: new Date(),
        submitInfo: {
            id: uuidv4(),
            updatedAt: new Date(),
            updatedByID: 'someone',
            updatedReason: 'submit',
            updatedBy: {
                id: 'someone',
                createdAt: new Date(),
                updatedAt: new Date(),
                givenName: 'Bob',
                familyName: 'Law',
                email: 'boblaw@example.com',
                role: 'STATE_USER',
                divisionAssignment: null,
                stateCode: stateCode,
            },
        },
        dsnpContract: null,
        unlockInfo: null,
        contractID: 'contractID',
        submitInfoID: null,
        unlockInfoID: null,
        programIDs: [statePrograms[0].id],
        populationCovered: 'MEDICAID' as const,
        submissionType: 'CONTRACT_ONLY' as const,
        riskBasedContract: false,
        submissionDescription: 'Test',
        stateContacts: [],
        supportingDocuments: [
            {
                id: uuidv4(),
                position: 0,
                contractRevisionID: 'contractRevisionID',
                createdAt: new Date(),
                updatedAt: new Date(),
                name: 'contract supporting doc',
                s3URL: 's3://bucketname/key/test1',
                sha256: '2342fwlkdmwvw',
                dateAdded: new Date(),
            },
            {
                id: uuidv4(),
                position: 1,
                contractRevisionID: 'contractRevisionID',
                createdAt: new Date(),
                updatedAt: new Date(),
                name: 'contract supporting doc 2',
                s3URL: 's3://bucketname/key/test1',
                sha256: '45662342fwlkdmwvw',
                dateAdded: new Date(),
            },
        ],
        contractType: 'BASE',
        contractExecutionStatus: 'EXECUTED',
        contractDocuments: [
            {
                id: uuidv4(),
                position: 0,
                contractRevisionID: 'contractRevisionID',
                createdAt: new Date(),
                updatedAt: new Date(),
                name: 'contract doc',
                s3URL: 's3://bucketname/key/test1',
                sha256: '8984234fwlkdmwvw',
                dateAdded: new Date(),
            },
        ],
        contractDateStart: new Date(Date.UTC(2025, 5, 1)),
        contractDateEnd: new Date(Date.UTC(2026, 4, 30)),
        managedCareEntities: ['MCO'],
        federalAuthorities: ['STATE_PLAN' as const],
        modifiedBenefitsProvided: false,
        modifiedGeoAreaServed: false,
        modifiedMedicaidBeneficiaries: false,
        modifiedRiskSharingStrategy: false,
        modifiedIncentiveArrangements: false,
        modifiedWitholdAgreements: false,
        modifiedStateDirectedPayments: false,
        modifiedPassThroughPayments: false,
        modifiedPaymentsForMentalDiseaseInstitutions: false,
        modifiedMedicalLossRatioStandards: false,
        modifiedOtherFinancialPaymentIncentive: false,
        modifiedEnrollmentProcess: false,
        modifiedGrevienceAndAppeal: false,
        modifiedNetworkAdequacyStandards: true,
        modifiedLengthOfContract: true,
        modifiedNonRiskPaymentArrangements: null,
        inLieuServicesAndSettings: null,
        statutoryRegulatoryAttestation: null,
        statutoryRegulatoryAttestationDescription: null,
        ...revision,
    }
}

export { mockInsertContractArgs, mockContractRevision }
