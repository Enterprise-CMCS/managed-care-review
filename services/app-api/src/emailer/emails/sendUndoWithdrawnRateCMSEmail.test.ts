import { testCMSUser, testStateUser } from '../../testHelpers/userHelpers'
import {
    mockContract,
    mockContractRev,
    mockMNState,
    mockRate,
    testEmailConfig,
    testStateAnalystsEmails,
} from '../../testHelpers/emailerHelpers'
import { sendUndoWithdrawnRateCMSEmail } from './sendUndoWithdrawnRateCMSEmail'
import type { RateType } from '../../domain-models'
import { mockRateRevision } from '../../testHelpers'

const testRate = mockRate({
    id: 'test-rate',
    createdAt: new Date('2025-03-05'),
    updatedAt: new Date('2025-03-05'),
    status: 'RESUBMITTED',
    reviewStatus: 'UNDER_REVIEW',
    consolidatedStatus: 'RESUBMITTED',
    parentContractID: 'parent-contract',
    withdrawnFromContracts: [],
    packageSubmissions: [
        {
            submitInfo: {
                updatedAt: new Date('2025-03-05'),
                updatedReason: 'Messed up - undoing withdraw',
                updatedBy: testStateUser({
                    email: 'undoer-of-withdraws@example.com',
                }),
            },
            submittedRevisions: [],
            rateRevision: {
                id: 'test-rate-revision',
                rateID: 'test-rate',
                submitInfo: {
                    updatedAt: new Date('2025-03-05'),
                    updatedReason: 'Messed up - undoing withdraw',
                    updatedBy: testCMSUser({
                        email: 'undoer-of-withdraws@example.com',
                    }),
                },
                createdAt: new Date('2025-03-05'),
                updatedAt: new Date('2025-03-05'),
                formData: {
                    rateType: 'NEW',
                    rateCapitationType: 'RATE_CELL',
                    rateDateStart: new Date('2025-03-05'),
                    rateDateEnd: new Date('2025-03-05'),
                    rateDateCertified: new Date('2025-03-05'),
                    rateCertificationName: 'test-rate-certification-name',
                    rateID: 'rate-id',
                    rateProgramIDs: [mockMNState().programs[0].id],
                    rateDocuments: [
                        {
                            s3URL: 's3://bucketname/key/test1',
                            name: 'ratedoc1.doc',
                            sha256: 'foobar',
                        },
                    ],
                    supportingDocuments: [],
                    certifyingActuaryContacts: [
                        {
                            name: 'Foo Person',
                            titleRole: 'Bar Job',
                            email: 'certifyingActuary@example.com',
                            actuarialFirm: 'GUIDEHOUSE',
                        },
                    ],
                    addtlActuaryContacts: [
                        {
                            name: 'Bar Person',
                            titleRole: 'Baz Job',
                            email: 'addtlActuaryContacts@example.com',
                            actuarialFirm: 'OTHER',
                            actuarialFirmOther: 'Some Firm',
                        },
                    ],
                    actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                },
            },
            contractRevisions: [
                mockContractRev({
                    submitInfo: {
                        updatedAt: new Date('2025-03-05'),
                        updatedReason: 'Messed up - undoing withdraw',
                        updatedBy: testCMSUser({
                            email: 'undoer-of-withdraws@example.com',
                        }),
                    },
                    formData: {
                        ...mockContractRev().formData,
                        stateContacts: [
                            {
                                name: 'parent-contract-state-contact',
                                titleRole: 'state contact',
                                email: 'parent-contract-state-contact@state.com',
                            },
                        ],
                    },
                }),
            ],
        },
    ],
    reviewStatusActions: [
        {
            rateID: 'test-rate',
            updatedBy: testCMSUser(),
            updatedReason: 'I am undoing a withdraw',
            actionType: 'UNDER_REVIEW',
            updatedAt: new Date('2025-03-05'),
        },
    ],
})

const statePrograms = mockMNState().programs
const stateAnalystEmails = () => [...testStateAnalystsEmails]

describe('sendUndoWithdrawnRateCMSEmail', () => {
    it('renders CMS for unwithdrawn rate as expected', async () => {
        const template = await sendUndoWithdrawnRateCMSEmail(
            testRate,
            statePrograms,
            stateAnalystEmails(),
            testEmailConfig()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template.toAddresses).toEqual(
            expect.arrayContaining([
                ...stateAnalystEmails(),
                ...testEmailConfig().dmcpSubmissionEmails,
                ...testEmailConfig().oactEmails,
            ])
        )
    })

    it('renders overall email for a withdrawn rate as expected', async () => {
        const template = await sendUndoWithdrawnRateCMSEmail(
            testRate,
            statePrograms,
            stateAnalystEmails(),
            testEmailConfig()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template.bodyHTML).toMatchSnapshot()
    })
})

describe('sendUndoWithdrawnRateCMSEmail error handling', () => {
    it('returns an error if rate consolidatedStatus is not RESUBMITTED', async () => {
        const rateWithWrongStatus = {
            ...testRate,
            consolidatedStatus: 'WITHDRAWN' as const,
        }
        const template = await sendUndoWithdrawnRateCMSEmail(
            rateWithWrongStatus,
            statePrograms,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Rate consolidated status should be RESUBMITTED'
        )
    })

    it('returns an error if the withdrawnFromContracts array was not emptied', async () => {
        const rateWithWrongStatus = {
            ...testRate,
            withdrawnFromContracts: [
                {
                    ...mockContract({
                        id: 'parent-contract',
                        stateNumber: 100,
                        packageSubmissions: [
                            {
                                submitInfo: {
                                    updatedAt: new Date('2025-03-05'),
                                    updatedReason:
                                        'Messed up - undoing withdraw',
                                    updatedBy: testCMSUser({
                                        email: 'undoer-of-withdraws@example.com',
                                    }),
                                },
                                submittedRevisions: [],
                                rateRevisions: [],
                                contractRevision: mockContractRev({
                                    submitInfo: {
                                        updatedAt: new Date('2025-03-05'),
                                        updatedReason:
                                            'Messed up - undoing withdraw',
                                        updatedBy: testCMSUser({
                                            email: 'undoer-of-withdraws@example.com',
                                        }),
                                    },
                                }),
                            },
                        ],
                    }),
                },
            ],
        }

        const template = await sendUndoWithdrawnRateCMSEmail(
            rateWithWrongStatus,
            statePrograms,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'withdrawnFromContracts should be empty'
        )
    })

    it('returns an error if there are no review status actions present', async () => {
        const missingReviewStatusActions = {
            ...testRate,
            reviewStatusActions: [],
        }

        const template = await sendUndoWithdrawnRateCMSEmail(
            missingReviewStatusActions,
            statePrograms,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Rate does not have any review status actions'
        )
    })

    it('returns an error if the latest review action is NOT "UNDER_REVIEW"', async () => {
        const rateWithWrongActionType = {
            ...testRate,
            reviewStatusActions: [
                {
                    rateID: 'test-rate',
                    updatedBy: testCMSUser(),
                    updatedReason: 'I must withdraw',
                    actionType: 'WITHDRAW' as const,
                    updatedAt: new Date('2025-03-10'),
                },
                {
                    rateID: 'test-rate',
                    updatedBy: testCMSUser(),
                    updatedReason: 'Test under review',
                    actionType: 'UNDER_REVIEW' as const,
                    updatedAt: new Date('2025-03-05'),
                },
                {
                    rateID: 'test-rate',
                    updatedBy: testCMSUser(),
                    updatedReason: 'Test withdraw',
                    actionType: 'WITHDRAW' as const,
                    updatedAt: new Date('2025-03-15'),
                },
            ],
        }

        const template = await sendUndoWithdrawnRateCMSEmail(
            rateWithWrongActionType,
            statePrograms,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Latest rate review action is not UNDER_REVIEW'
        )
    })

    it('returns an error if there is no rateCertificationName', async () => {
        const rateWithoutRateCertificationName: RateType = {
            ...testRate,
            packageSubmissions: [
                {
                    ...testRate.packageSubmissions[0],
                    rateRevision: {
                        ...mockRateRevision(),
                        id: 'test-rate-revision',
                        rateID: 'test-rate',
                        submitInfo: {
                            updatedAt: new Date('2025-03-05'),
                            updatedReason: 'Messed up - undoing withdraw',
                            updatedBy: testCMSUser({
                                email: 'undoer-of-withdraws@example.com',
                            }),
                        },
                        unlockInfo: undefined,
                        createdAt: new Date('2025-03-05'),
                        updatedAt: new Date('2025-03-05'),
                        formData: {
                            rateType: 'NEW',
                            rateCapitationType: 'RATE_CELL',
                            rateDateStart: new Date('2025-03-05'),
                            rateDateEnd: new Date('2025-03-05'),
                            rateDateCertified: new Date('2025-03-05'),
                            rateCertificationName: '',
                            rateID: 'test-rate',
                            rateProgramIDs: [mockMNState().programs[0].id],
                            rateDocuments: [
                                {
                                    s3URL: 's3://bucketname/key/test1',
                                    name: 'ratedoc1.doc',
                                    sha256: 'foobar',
                                },
                            ],
                            supportingDocuments: [],
                            certifyingActuaryContacts: [
                                {
                                    name: 'Foo Person',
                                    titleRole: 'Bar Job',
                                    email: 'certifyingActuary@example.com',
                                    actuarialFirm: 'GUIDEHOUSE',
                                },
                            ],
                            addtlActuaryContacts: [
                                {
                                    name: 'Bar Person',
                                    titleRole: 'Baz Job',
                                    email: 'addtlActuaryContacts@example.com',
                                    actuarialFirm: 'OTHER',
                                    actuarialFirmOther: 'Some Firm',
                                },
                            ],
                            actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                        },
                    },
                },
            ],
        }

        const template = await sendUndoWithdrawnRateCMSEmail(
            rateWithoutRateCertificationName,
            statePrograms,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Error parsing for rateCertificationName'
        )
    })

    it('returns an error if there is no rateID', async () => {
        const rateWithoutRateID: RateType = {
            ...testRate,
            packageSubmissions: [
                {
                    ...testRate.packageSubmissions[0],
                    rateRevision: {
                        id: 'test-rate-revision',
                        rateID: 'test-rate',
                        submitInfo: {
                            updatedAt: new Date('2025-03-05'),
                            updatedReason: 'Messed up - undoing withdraw',
                            updatedBy: testCMSUser({
                                email: 'undoer-of-withdraws@example.com',
                            }),
                        },
                        createdAt: new Date('2025-03-05'),
                        updatedAt: new Date('2025-03-05'),
                        formData: {
                            rateType: 'NEW',
                            rateCapitationType: 'RATE_CELL',
                            rateDateStart: new Date('2025-03-05'),
                            rateDateEnd: new Date('2025-03-05'),
                            rateDateCertified: new Date('2025-03-05'),
                            rateCertificationName: 'test-rate',
                            rateID: '',
                            rateProgramIDs: [mockMNState().programs[0].id],
                            rateDocuments: [
                                {
                                    s3URL: 's3://bucketname/key/test1',
                                    name: 'ratedoc1.doc',
                                    sha256: 'foobar',
                                },
                            ],
                            supportingDocuments: [],
                            certifyingActuaryContacts: [
                                {
                                    name: 'Foo Person',
                                    titleRole: 'Bar Job',
                                    email: 'certifyingActuary@example.com',
                                    actuarialFirm: 'GUIDEHOUSE',
                                },
                            ],
                            addtlActuaryContacts: [
                                {
                                    name: 'Bar Person',
                                    titleRole: 'Baz Job',
                                    email: 'addtlActuaryContacts@example.com',
                                    actuarialFirm: 'OTHER',
                                    actuarialFirmOther: 'Some Firm',
                                },
                            ],
                            actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                        },
                    },
                },
            ],
        }

        const template = await sendUndoWithdrawnRateCMSEmail(
            rateWithoutRateID,
            statePrograms,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe('Error parsing for rateID')
    })

    it('returns an error if there are no contractRevisions', async () => {
        const rateWithoutContractRevisions: RateType = {
            ...testRate,
            packageSubmissions: [
                {
                    ...testRate.packageSubmissions[0],
                    rateRevision: {
                        id: 'test-rate-revision',
                        rateID: 'test-rate',
                        submitInfo: {
                            updatedAt: new Date('2025-03-05'),
                            updatedReason: 'Messed up - undoing withdraw',
                            updatedBy: testCMSUser({
                                email: 'undoer-of-withdraws@example.com',
                            }),
                        },
                        createdAt: new Date('2025-03-05'),
                        updatedAt: new Date('2025-03-05'),
                        formData: {
                            rateType: 'NEW',
                            rateCapitationType: 'RATE_CELL',
                            rateDateStart: new Date('2025-03-05'),
                            rateDateEnd: new Date('2025-03-05'),
                            rateDateCertified: new Date('2025-03-05'),
                            rateCertificationName:
                                'test-rate-certification-name',
                            rateID: 'test-rate-id',
                            rateProgramIDs: [mockMNState().programs[0].id],
                            rateDocuments: [
                                {
                                    s3URL: 's3://bucketname/key/test1',
                                    name: 'ratedoc1.doc',
                                    sha256: 'foobar',
                                },
                            ],
                            supportingDocuments: [],
                            certifyingActuaryContacts: [
                                {
                                    name: 'Foo Person',
                                    titleRole: 'Bar Job',
                                    email: 'certifyingActuary@example.com',
                                    actuarialFirm: 'GUIDEHOUSE',
                                },
                            ],
                            addtlActuaryContacts: [
                                {
                                    name: 'Bar Person',
                                    titleRole: 'Baz Job',
                                    email: 'addtlActuaryContacts@example.com',
                                    actuarialFirm: 'OTHER',
                                    actuarialFirmOther: 'Some Firm',
                                },
                            ],
                            actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                        },
                    },
                    contractRevisions: [],
                },
            ],
        }

        const template = await sendUndoWithdrawnRateCMSEmail(
            rateWithoutContractRevisions,
            statePrograms,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Rate does not have any contract revisions'
        )
    })
})
