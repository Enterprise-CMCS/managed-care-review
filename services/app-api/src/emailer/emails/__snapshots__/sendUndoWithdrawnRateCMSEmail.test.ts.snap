// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sendUndoWithdrawnRateCMSEmail > renders overall email for a withdrawn rate as expected 1`] = `
"<h1 style="font-weight: normal;">Rate test-rate-certification-name status was updated to 'Submitted' by CMS</h1>
<b>Updated by:</b> zuko@example.com<br />
<b>Updated on:</b> 03/04/2025<br />
<b>Reason:</b> I am undoing a withdraw<br />
<b>Related contract action(s):</b>
<a href="http://localhost/submissions/12345">MCR-MN-0003-SNBC</a><br />
<br />
"
`;
